src/loki/polygon_search.cc

```diff
@@ -0,0 +1,115 @@
+#include "loki/polygon_search.h"
+
+namespace bg = boost::geometry;
+namespace vm = valhalla::midgard;
+namespace vb = valhalla::baldr;
+
+namespace valhalla {
+namespace loki {
+
+std::unordered_set<vb::GraphId> edges_in_rings(const std::vector<ring_bg_t>& rings,
+                                               baldr::GraphReader& reader,
+                                               const std::shared_ptr<sif::DynamicCost>& costing) {
+  // Get the lowest level and tiles
+  const auto tiles = vb::TileHierarchy::levels().back().tiles;
+  const auto bin_level = vb::TileHierarchy::levels().back().level;
+
+  // keep track which tile's bins intersect which rings
+  bins_collector bins_intersected;
+  std::unordered_set<vb::GraphId> avoid_edge_ids;
+
+  // first pull out all *unique* bins which intersect the rings
+  for (size_t ring_idx = 0; ring_idx < rings.size(); ring_idx++) {
+    auto ring = rings[ring_idx];
+    auto line_intersected = tiles.Intersect(ring);
+    for (const auto& tb : line_intersected) {
+      for (const auto& b : tb.second) {
+        bins_intersected[tb.first][b].push_back(ring_idx);
+      }
+    }
+  }
+
+  std::vector<int32_t> tile_keys(bins_intersected.size());
+  std::for_each(bins_intersected.begin(), bins_intersected.end(),
+                [&tile_keys](
+                    const std::pair<int32_t, std::unordered_map<unsigned short, std::vector<size_t>>>&
+                        e) { tile_keys.push_back(e.first); });
+  std::sort(tile_keys.begin(), tile_keys.end());
+  for (const auto& tile_id : tile_keys) {
+    auto& tile_bins = bins_intersected[tile_id];
+    auto tile = reader.GetGraphTile({static_cast<uint32_t>(tile_id), bin_level, 0});
+    if (!tile) {
+      continue;
+    }
+    for (const auto& bin : tile_bins) {
+      // tile will be mutated most likely in the loop
+      if (tile->id().tileid() != tile_id) {
+        tile = reader.GetGraphTile({static_cast<uint32_t>(tile_id), bin_level, 0});
+      }
+      for (const auto& edge_id : tile->GetBin(bin.first)) {
+        // weed out duplicates early on
+        // TODO: is this really worth it?
+        if (avoid_edge_ids.find(edge_id) != avoid_edge_ids.end()) {
+          continue;
+        }
+        // TODO: optimize the tile switching by enqueuing edges
+        // from other levels & tiles and process them after this big loop
+        if (edge_id.Tile_Base() != tile->header()->graphid().Tile_Base() &&
+            !reader.GetGraphTile(edge_id, tile)) {
+          continue;
+        }
```

Comment:
not sure how far one can optimise here.. as then the same optimization might be applied to getting the opp edges?